uavcan/node/SemanticVersion.1.0.uavcan

```diff
@@ -0,0 +1,7 @@
+#
```

Comment:
If this is going to be used on software as well patch version is relevant. What about:

```
uint6 major
uint5 minor
uint5 patch
```

Reply:
We use semantic version tuples for the protocol itself, hardware, software, and data type definitions. The patch version can be sensibly used only for software. Do we really need to add a special case for that, given that there are other means of identifying a particular build of software? Namely, the version control system revision ID and the image hash.

Reply:
Well, if we're going to call it SemVer we should have the ability to use a patch field. If we're going to use it for software it would be nice to have the patch field as well. I wouldn't even call it a "special case".

The alternative definition still uses 2 bytes and I think we can be quite comfortable with the ranges. I don't think we're worse off in any way with this alternative definition.

We can consider switching minor to the 6-bit field.

Reply:
What I don't like here is that we would have an unused patch number in the other three use cases where it doesn't serve much purpose. SemVer was created for software, not vehicular systems, so maybe it's just not the best option here. How about any of the following options:

* We stop calling it `SemanticVersion` and call it just `MajorMinorVersion` or something else as long as it's generic. In the data type documentation we can explain how it should be used. References to semver.org are optional.
* We define another, abridged version struct alongside `SemanticVersion`, and use it for everything except software version. The software version will be represented by `SemanticVersion` with the patch number. The downside is that it's a new data type.

I am still not sure if the patch number is required because release management of embedded software is usually different compared to computer software: updates are always risky and expensive. I don't think the patch number can be sensibly applied to safety-critical systems, which are usually not updated unless there are major issues.

Reply:
> > we would have an unused patch number

> Not really unused, it just happens to always be 0.

Which means that it is unused.

> A lot of security patches (major issues) doesn't change the external interface in any way.

It's absolutely irrelevant. In high safety assurance applications the very act of updating the firmware, even to the very same version down to the same binary image, can be a noticeable risk, significant enough to avoid that unless absolutely necessary. You don't even have to go that far, consider this: unless you are using a certified compiler or object code traceability, just a recompilation of the exact same codebase adds some amount of uncertainty and therefore risk.

> At the end of the day, I don't see much problem just putting a zero for the patch version.

The problem is that we're adding a new entity and not buying anything useful by doing that. With the existing API we can already do the following:

* Identifying the major and the minor versions of software, hardware, protocol, and each data type.
* Determining whether two deployed software images are of the exact same version via the image CRC. The patch version doesn't help here as it doesn't take into account the compilation process and ultimately depends on the developer's assessment of the impact of their changes, which can be unreliable.
* Mapping from the deployed software version to the corresponding revision of the source code via the VCS revision identifier.

Adding a new union for either major-minor or semver doesn't make sense because it adds more variance where it's unnecessary.

If a semantic version can't be used without the patch number, we can rename the type `SemanticVersion` to `MajorMinorVersion` (or similar) and keep everything else intact.

Reply:
The rename route seems sensible, I'm convinced. As you suggested we can call it `MajorMinorVersion` or even just `Version` since this is how versions are used in DSDL. :+1: